File: go-algo/collections/min_stack.go
package collections


type MinStack struct {
	main, min []int
}

func NewMinStack() MinStack {
	return MinStack{[]int{}, []int{}}
}

func (ms *MinStack) Push(x int)  {
	ms.main = append(ms.main, x)
	if len(ms.min) == 0 || x <= ms.min[len(ms.min)-1] {
		ms.min = append(ms.min, x)
	}
}

func (ms *MinStack) Pop()  {
	n, n2 := len(ms.main)-1, len(ms.min)-1
	if ms.min[n2] == ms.main[n] {
		ms.min = ms.min[:n2]
	}
	ms.main = ms.main[:n]
}

func (ms *MinStack) Top() int {
	return ms.main[len(ms.main)-1]
}

func (ms *MinStack) GetMin() int {
	return ms.min[len(ms.min)-1]
}

File: py-algo/leetcode/design/min-stack.py
class MinStack:
    def __init__(self):
        self.min_stk = []
        self.stk = []

    # @param x, an integer
    def push(self, x):
        self.stk.append(x)
        if not self.min_stk or x < self.min_stk[-1]:
            self.min_stk.append(x)

    # @return nothing
    def pop(self):
        if self.stk:
            val = self.stk.pop()
            if val == self.min_stk[-1]:
                self.min_stk.pop()
            return val

    # @return an integer
    def top(self):
        return self.stk[-1] if self.stk else -1

    # @return an integer
    def getMin(self):
        return self.min_stk[-1] if self.min_stk else -1